### PR TITLE
network: wait for DSA master up

### LINKED ISSUE #7478: systemd-networkd support for DSA ("Distributed Switch Architecture")
### Submission type

 - Bug report

### systemd version the issue has been seen with

235

### Used distribution

> Debian

### In case of bug report: Expected behaviour you didn't see

> Interface `wan` (which is linked to `eth0`) comes up

### In case of bug report: Unexpected behaviour you saw

> well, it doesn't

### In case of bug report: Steps to reproduce the problem

Networking on the EspressoBin is special: there's an internal switch. The kernel comes up with this setup:

```
# ip link
[…]
4: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 532
    link/ether 00:00:00:00:51:81 brd ff:ff:ff:ff:ff:ff
5: wan@eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:00:00:00:51:81 brd ff:ff:ff:ff:ff:ff
6: lan0@eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:00:00:00:51:81 brd ff:ff:ff:ff:ff:ff
7: lan1@eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:00:00:00:51:81 brd ff:ff:ff:ff:ff:ff
```
(except that initially eth0 is down)

I have two `systemd.network` files for `eth0` and `wan`:
```
# cd /etc/systemd/network; grep . *
49-switch.network:[Match]
49-switch.network:Name=eth0
49-switch.network:[Network]
49-switch.network:DHCP=no
50-main.network:[Match]
50-main.network:Name=wan
50-main.network:[Network]
50-main.network:DHCP=no
50-main.network:[Address]
50-main.network:Address=10.107.0.1/24
```
However, depending on the Gods of Random Chance, I need to restart `systemd-networkd` in order to bring up the `wan` interface. The first start only enables `eth0`.

Workaround:
```
# cat /etc/systemd/system/systemd-networkd.service.d/after-udev.conf 
[Service]
ExecStartPre=/sbin/ip link set dev eth0 up
```

### PR DESCRIPTION
**Pull request by yuwata**
This is for the DSA subsystem, which have several stacked interfaces
on the master interface. To bring up a stacked interface, it is necessary
that the master is already up. See
https://github.com/systemd/systemd/issues/7478#issuecomment-348508263.

Note this is not necessary for newer kernels which includes
https://github.com/torvalds/linux/commit/9d5ef190e5615a7b63af89f88c4106a5bc127974.

Fixes #7478.

### CODE REVIEW

**Comment by yuwata** [comment]
The centos CI failure is #21675, unrelated.

**Review by aryuemaan — APPROVED** [review]

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -2550,6 +2550,23 @@ static int link_new(Manager *manager, sd_netlink_message *message, Link **ret) {
...
+                        dsa_master_ifindex = 0;
+                else if (r < 0)
+                        log_link_warning_errno(link, r, "rtnl: failed to read ifindex of the DSA master interface, ignoring: %m");
+                else if (dsa_master_ifindex > INT_MAX) {
+                        log_link_warning(link, "rtnl: received too large DSA master ifindex (%"PRIu32" > INT_MAX), ignoring.",
+                                         dsa_master_ifindex);
+                        dsa_master_ifindex = 0;
+                }
+
+                link->dsa_master_ifindex = (int) dsa_master_ifindex;
```
Comment (bluca) [inline src/network/networkd-link.c:2567]: isn't `dsa_master_ifindex` uninitialized if it goes through the second branch of the if statement?

File: src/network/networkd-setlink.c [hunk 2]
```diff
@@ -958,7 +993,9 @@ static int link_up_or_down_handler_internal(sd_netlink *rtnl, sd_netlink_message
                 goto on_error;
 
         r = sd_netlink_message_get_errno(m);
-        if (r < 0) {
+        if (r == -ENETDOWN && up && link_up_dsa_slave(link) > 0)
```
Comment (bluca) [inline src/network/networkd-setlink.c:996]: why is it printing the error when `link_up_dsa_slave` returns success?

File: src/network/networkd-setlink.c [hunk 3]
```diff
@@ -1130,9 +1167,21 @@ int link_request_to_activate(Link *link) {
...
+static bool link_is_ready_to_bring_up_or_down(Link *link, bool up) {
         assert(link);
 
+        if (up && link->dsa_master_ifindex > 0) {
+                Link *master;
+
+                /* The master inteface must be up. See comments in link_up_dsa_slave(). */
+
+                if (link_get_by_index(link->manager, link->dsa_master_ifindex, &master) < 0)
+                        false;
```
Comment (bluca) [inline src/network/networkd-setlink.c]: did you mean `return false` here and below?

File: src/network/networkd-setlink.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-setlink.c:996]: This intends to log the first attempt is failed in log level. The log message below is slightly updated.

**Pull request merged**